---
title: "Step 0: Pre-processing with all volunteers"
author: "Carlos Gallardo & Christian Oertlin"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    highlight: tango
---

```{r setup, include=FALSE}
# Set path
path <- strsplit(dirname(rstudioapi::getActiveDocumentContext()$path), '/')[[1]]
path <- path[-c(length(path))]
path <- paste(path, collapse = '/')
knitr::opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo = T, message = F, warning = F, dev = 'cairo_pdf')
def.chunk.hook  <- knitr::knit_hooks$get("chunk")
knitr::knit_hooks$set(chunk = function(x, options) {
  size <- 'scriptsize'
  x <- def.chunk.hook(x, options)
  paste0("\n \\", size, "\n\n", x, "\n\n \\normalsize")
})

```

```{r}
# Import libraries and helper functions
source("code/helper_functions.R")
library(tidyverse)
library(magrittr)
library(patchwork)
library(DESeq2)
library(limma)
library(variancePartition)
library(RColorBrewer)
library(ComplexHeatmap)

# Colors
colPals <- vector(mode = "list")
colPals$time <- setNames(c("#FBAA3E", "#2C83BE", "#3EB6BD", "#A3D5B3", "#CD71A8"),
                         nm = c("day0", "day7", "day14", "day21", "day28"))
colPals$time_light <- setNames(c("#FDD6A1", "#A2CDE9", "#AFE2E5", "#DDF0E3", "#E8BDD6"),
                               nm = c("day0", "day7", "day14", "day21", "day28"))
colPals$time_dark <- setNames(c("#D87E04", "#174564", "#1F5C60", "#49A065", "#AA3C7E"),
                              nm = c("day0", "day7", "day14", "day21", "day28"))
colPals$inferno <- c("#000004", "#420A68", "#932667", "#DD513A", "#FCA50A", "#FCFFA4")
colPals$factors <- setNames(c('#E31D27','#FA9F1C','#9A509F','#C1C1C1'),
                            nm = c("volunteer","time","batch","Residuals"))

```


# Important note:

> This script contains the exploratory analysis conducted on all 10 volunteers participating 
> in the dry immersion (DI) study. As shown below, v9 and v10 specifically show elevated 
> expression of monocyte genes (e.g., CDKN1A, NR4A1, NR4A2, MYADM, IRS2 and CD83) suggesting
> contamination for these samples. As per our exclusion criteria we chose to remove these
> for further analyses. The data provided includes volunteers 1 to 8 which are used for 
> investigation in the study. Data with v9 and v10 can be provided upon request to reproduce
> the pre-processing steps in this script.


# Load data

## Background annotation

```{r}
ann_data <- read.table(
  file = 'data/resources/gene_annotation_ensembl_v104.txt',
  stringsAsFactors = FALSE,
  sep = "\t",
  header = TRUE,
  fill = FALSE,
  quote = "") %>% 
  dplyr::rename(Geneid = ensembl_gene_id)

```


## RNAseq expression data

```{r}
RNAseq <- vector("list")

RNAseq[["unfilt"]][["rawdata"]] <- read.table(file = 'data/rnaseq/exon_counts_all.txt',
                                              stringsAsFactors = FALSE,
                                              sep = "\t",
                                              header = TRUE)

# Make duplicated gene names unique
RNAseq[["unfilt"]][["rawdata"]] <- RNAseq[["unfilt"]][["rawdata"]] %>% 
  mutate(GeneSymbol = uniquify(plyr::mapvalues(.$Geneid,
                                               from = ann_data$Geneid,
                                               to = ann_data$external_gene_name,
                                               warn_missing = F), sep = '_'))

RNAseq[["unfilt"]][["annotation"]] <- RNAseq[["unfilt"]][["rawdata"]] %>% 
  select(Geneid, GeneSymbol) %>% 
  inner_join(ann_data, by = "Geneid") %>% 
  select(-external_gene_name)

RNAseq[["unfilt"]][["design"]] <- read.table(file = 'data/RNAseq/design_mtx_all.txt',
                                             stringsAsFactors = FALSE,
                                             sep = "\t",
                                             header = TRUE) %>% 
  mutate(sample = factor(sample, levels = sample),
         batch = factor(batch, levels = unique(batch)),
         volunteer = factor(volunteer, levels = unique(volunteer)),
         time = factor(time, levels = unique(time)))
                                   
RNAseq[["unfilt"]][["counts"]] <- RNAseq[["unfilt"]][["rawdata"]] %>% 
  select(-c(1:3)) %>% 
  column_to_rownames("GeneSymbol")

```


# Pre-processing

## Filtering zero count genes

```{r}
paste("Raw feature count:", nrow(RNAseq$unfilt$counts))
tokeep <- rowSums(RNAseq$unfilt$counts) > 0
paste("Non-zero feature count:", sum(tokeep))
RNAseq$unfilt$rawdata <- RNAseq$unfilt$rawdata[tokeep,]
RNAseq$unfilt$annotation <- RNAseq$unfilt$annotation[tokeep,]
RNAseq$unfilt$counts <- RNAseq$unfilt$counts[tokeep,]
rm(tokeep)

```


## Library sizes and gene expression distributions

```{r, fig.height=5, fig.width=12}
df <- RNAseq$unfilt$design
df$lib.size <- colSums(RNAseq$unfilt$counts)

ggplot(df, aes(x=sample, y=lib.size/1e6, fill=volunteer)) +
  geom_bar(stat = "identity", width = 0.8) +
  xlab("") +
  ylab("Library size (million reads)") +
  scale_x_discrete(expand =expansion(mult = c(.02, .02))) +
  scale_y_continuous(expand =expansion(mult = c(.02, .05))) +
  theme_custom(
    axis.text.x.bottom = element_text(angle = 90, hjust = 1, vjust = 0.3)
  )

```

```{r, fig.height=6, fig.width=12}
df <- RNAseq$unfilt$counts %>% 
  rownames_to_column(var = "geneID") %>% 
  pivot_longer(cols = c(2:length(.)),
               names_to = "sample") %>% 
  dplyr::rename(counts = value)

df$volunteer <- rep(RNAseq$unfilt$design$volunteer, dim(RNAseq$unfilt$counts)[1])
df$time <- rep(RNAseq$unfilt$design$time, dim(RNAseq$unfilt$counts)[1])
df$sample <- factor(df$sample, levels = names(RNAseq$unfilt$counts))

ggplot(df, aes(x=sample, y=log2(counts+1), fill=volunteer)) +
  geom_violin(scale = "area") +
  xlab("") +
  ylab(expression(Log[2](counts+1))) +
  scale_x_discrete(expand =expansion(mult = c(.02, .02))) +
  scale_y_continuous(expand =expansion(mult = c(.02, .05))) +
  theme_custom(
    axis.text.x.bottom = element_text(angle = 90, hjust = 1, vjust = 0.3),
    legend.position = "top"
  )

```


## Filtering low abundance genes

```{r, fig.height=5, fig.width=9}
# Calculate CPM
RNAseq[["unfilt"]][["cpm"]] <- cpm.normalize(RNAseq$unfilt$counts)

abovethresh <- RNAseq$unfilt$cpm > 1

df <- data.frame(samples = factor(seq(0,ncol(RNAseq$unfilt$cpm),1), 
                                  levels = rev(seq(0,ncol(RNAseq$unfilt$cpm),1))),
                 genes = c(table(rowSums(abovethresh)))) %>% 
  mutate(cumulative = rev(cumsum(rev(genes)))) %>% 
  mutate(remaining = sum(genes)-cumulative) %>% 
  pivot_longer(cols = c("cumulative","remaining"),
               names_to = "group") %>% 
  mutate(group = factor(group, levels = c("remaining", "cumulative")))

ggplot(data=df, aes(x=samples, y=value, fill=group)) +
  geom_bar(color="black", size=0.5, width=0.8, position="stack", stat="identity") +
  geom_hline(yintercept = unlist(df[df$samples == "20" & df$group == "cumulative", "value"]),
             linetype="solid", size=1, color="#EF2126") +
  geom_vline(xintercept = "20", linetype="solid", size=1, color="#EF2126") +
  xlab("# of samples present (CPM>1)") +
  ylab("# of genes") +
  scale_x_discrete(expand =expansion(mult = c(.02, .02))) +
  scale_y_continuous(expand =expansion(mult = c(.02, .00)),
                     limits = c(0,45000), breaks = seq(0,45000,15000)) +
  scale_fill_manual(values = c("grey80", "grey20")) +
  theme_custom(base_size = 8)

```

```{r}
tokeep <- rowSums(abovethresh) >= 20
paste("Pre-filtering gene count:", length(tokeep))
paste("Genes below abundance threshold:", length(tokeep)-sum(tokeep))
paste("Remaining genes:", sum(tokeep))

```

```{r}
# Filter genes
RNAseq[["filt"]][["rawdata"]] <- RNAseq$unfilt$rawdata[tokeep,]
RNAseq[["filt"]][["annotation"]] <- RNAseq$unfilt$annotation[tokeep,]
RNAseq[["filt"]][["design"]] <- RNAseq$unfilt$design
RNAseq[["filt"]][["counts"]] <- RNAseq$unfilt$counts[tokeep,]

# Normalize
RNAseq[["filt"]][["cpm"]] <- cpm.normalize(RNAseq$filt$counts)

```


## Gene expression distribution post-filtering

```{r, fig.height=6, fig.width=12}
# Expression distribution post-filtering
df <- RNAseq$filt$counts %>% 
  rownames_to_column(var = "geneID") %>% 
  pivot_longer(cols = c(2:length(.)),
               names_to = "sample") %>% 
  dplyr::rename(counts = value)

df$volunteer <- rep(RNAseq$filt$design$volunteer, dim(RNAseq$filt$counts)[1])
df$time <- rep(RNAseq$filt$design$time, dim(RNAseq$filt$counts)[1])
df$sample <- factor(df$sample, levels = names(RNAseq$filt$counts))

ggplot(df, aes(x=sample, y=log2(counts+1), fill=volunteer)) +
  geom_violin(scale = "area") +
  xlab("") +
  ylab(expression(Log[2](counts+1))) +
  scale_x_discrete(expand =expansion(mult = c(.02, .02))) +
  scale_y_continuous(expand =expansion(mult = c(.02, .05))) +
  theme_custom(
    axis.text.x.bottom = element_text(angle = 90, hjust = 1, vjust = 0.3),
    legend.position = "top"
  )

```


## Normalized gene expression distribution post-filtering

```{r, fig.height=6, fig.width=12}
# Normalized expression distribution post-filtering
df <- RNAseq$filt$counts %>% 
  rownames_to_column(var = "geneID") %>% 
  pivot_longer(cols = c(2:length(.)),
               names_to = "sample") %>% 
  dplyr::rename(counts = value)

df$volunteer <- rep(RNAseq$filt$design$volunteer, dim(RNAseq$filt$counts)[1])
df$time <- rep(RNAseq$filt$design$time, dim(RNAseq$filt$counts)[1])
df$sample <- factor(df$sample, levels = names(RNAseq$filt$counts))

df$cpm <- RNAseq$filt$cpm %>% 
  pivot_longer(cols = c(1:length(.)),
               names_to = "sample") %>% 
  select(value) %>% 
  unlist()

ggplot(df, aes(x=sample, y=log2(cpm+1), fill=volunteer)) +
  geom_violin(scale = "area") +
  xlab("") +
  ylab(expression(Log[2](CPM+1))) +
  scale_x_discrete(expand =expansion(mult = c(.02, .02))) +
  scale_y_continuous(expand =expansion(mult = c(.02, .05))) +
  theme_custom(
    axis.text.x.bottom = element_text(angle = 90, hjust = 1, vjust = 0.3),
    legend.position = "top"
  )

```


# Transcriptome differences

## Sample correlations

```{r, fig.height=9, fig.width=10}
df <- cor(log2(RNAseq$filt$cpm+1), method = "spearman") %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "sample1") %>% 
  mutate(across(everything(), as.character)) %>% 
  pivot_longer(cols = c(2:length(.)),
               names_to = "sample2") %>% 
  dplyr::rename(r = value) %>% 
  mutate(sample1 = factor(sample1, levels = names(RNAseq$filt$counts)),
         sample2 = factor(sample2, levels = names(RNAseq$filt$counts)),
         r = as.numeric(r))

ggplot(df, aes(x=sample1, y=sample2, fill= r)) + 
  geom_tile() +
  scale_y_discrete(limits=rev) +
  scale_fill_gradientn(colours = rev(colPals$inferno)) +
  theme_custom(
    axis.text.x.bottom = element_text(angle = 90, hjust = 1, vjust = 0.3),
    legend.position = "right",
    legend.justification = "top"
  )

```


## DESeq2 analysis

```{r}
# DESeq2 pipeline
dsData <- DESeqDataSetFromMatrix(countData = RNAseq$filt$counts,
                                 colData = RNAseq$filt$design,
                                 design = ~volunteer + time)
dsData <- estimateSizeFactors(dsData)
dsData <- DESeq(dsData, test = "LRT", reduced = ~volunteer)

RNAseq$filt[["DESeq_norm"]] <- counts(dsData, normalized=TRUE) %>% as.data.frame()
RNAseq$filt[["DESeq_vst"]] <- assay(vst(dsData, blind=FALSE)) %>% as.data.frame()
RNAseq$filt[["DESeq_rlog"]] <- assay(rlog(dsData, blind=FALSE)) %>% as.data.frame()

DESeq2_DEGs <- list(
  GLMtime = results(dsData, test = "LRT", independentFiltering = T),
  day7Vsday0 = results(dsData, contrast=c("time", "day7", "day0"), test = "Wald", independentFiltering = T),
  day14Vsday0 = results(dsData, contrast=c("time", "day14", "day0"), test = "Wald", independentFiltering = T),
  day21Vsday0 = results(dsData, contrast=c("time", "day21", "day0"), test = "Wald", independentFiltering = T),
  day28Vsday0 = results(dsData, contrast=c("time", "day28", "day0"), test = "Wald", independentFiltering = T)
)
DESeq2_DEGs <- lapply(DESeq2_DEGs, as.data.frame)
DESeq2_DEGs$GLMtime <- DESeq2_DEGs$GLMtime %>% 
  add_column(logFC_day7Vsday0 = DESeq2_DEGs$day7Vsday0$log2FoldChange, .before = "log2FoldChange") %>% 
  add_column(logFC_day14Vsday0 = DESeq2_DEGs$day14Vsday0$log2FoldChange, .before = "log2FoldChange") %>% 
  add_column(logFC_day21Vsday0 = DESeq2_DEGs$day21Vsday0$log2FoldChange, .before = "log2FoldChange") %>% 
  add_column(logFC_day28Vsday0 = DESeq2_DEGs$day28Vsday0$log2FoldChange, .before = "log2FoldChange") %>% 
  select(-log2FoldChange)
DESeq2_DEGs <- lapply(DESeq2_DEGs, function(x) mutate(x, padj=ifelse(is.na(padj), 1, padj)))
DESeq2_DEGs <- lapply(DESeq2_DEGs, function(x) arrange(x, padj))
DESeq2_DEGs <- lapply(DESeq2_DEGs, rownames_to_column, var = "GeneSymbol")
DESeq2_DEGs <- lapply(DESeq2_DEGs, inner_join, y = RNAseq$filt$annotation, by = "GeneSymbol")

DESeq2_DEGs_filt <- list(
  padj_02 = lapply(DESeq2_DEGs, function(x) x %>% filter(padj<0.2)),
  padj_005 = lapply(DESeq2_DEGs, function(x) x %>% filter(padj<0.05)),
  padj_001 = lapply(DESeq2_DEGs, function(x) x %>% filter(padj<0.01))
)

```


## Batch correction

```{r, fig.height=6, fig.width=12}
# Remove batch effects from data
design_mtx <- model.matrix(~time, data = RNAseq$filt$design)
RNAseq[["filt"]][["DESeq_vst_nobatch"]] <- limma::removeBatchEffect(RNAseq$filt$DESeq_vst,
                                                                    batch = dsData$volunteer,
                                                                    design = design_mtx) %>% as.data.frame()
RNAseq[["filt"]][["DESeq_rlog_nobatch"]] <- limma::removeBatchEffect(RNAseq$filt$DESeq_rlog,
                                                                     batch = dsData$volunteer,
                                                                     design = design_mtx) %>% as.data.frame()

# Plot sample correlations
df <- cor(RNAseq$filt$DESeq_vst, method = "spearman") %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "sample1") %>% 
  mutate(across(everything(), as.character)) %>% 
  pivot_longer(cols = c(2:length(.)),
               names_to = "sample2") %>% 
  dplyr::rename(r = value) %>% 
  mutate(sample1 = factor(sample1, levels = names(RNAseq$filt$counts)),
         sample2 = factor(sample2, levels = names(RNAseq$filt$counts)),
         r = as.numeric(r))

p1 <- ggplot(df, aes(x=sample1, y=sample2, fill= r)) + 
  geom_tile() +
  scale_x_discrete(labels=paste(RNAseq$filt$design$volunteer, 
                                RNAseq$filt$design$time, 
                                sep = '_')) +
  scale_y_discrete(limits=rev, labels=rev(paste(RNAseq$filt$design$volunteer, 
                                                RNAseq$filt$design$time, 
                                                sep = '_'))) +
  scale_fill_gradientn(colours = rev(colPals$inferno)) +
  xlab('') +
  ylab('') +
  ggtitle('Normalized') +
  theme_custom(
    base_size = 6,
    axis.text.x.bottom = element_text(angle = 90, hjust = 1, vjust = 0.3),
    legend.position = "none",
    plot.title = element_text(size=16, face='bold', hjust=0.5)
  )

df2 <- cor(RNAseq$filt$DESeq_vst_nobatch, method = "spearman") %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "sample1") %>% 
  mutate(across(everything(), as.character)) %>% 
  pivot_longer(cols = c(2:length(.)),
               names_to = "sample2") %>% 
  dplyr::rename(r = value) %>% 
  mutate(sample1 = factor(sample1, levels = names(RNAseq$filt$counts)),
         sample2 = factor(sample2, levels = names(RNAseq$filt$counts)),
         r = as.numeric(r))

p2 <- ggplot(df2, aes(x=sample1, y=sample2, fill= r)) + 
  geom_tile() +
  scale_x_discrete(labels=paste(RNAseq$filt$design$volunteer, 
                                RNAseq$filt$design$time, 
                                sep = '_')) +
  scale_y_discrete(limits=rev, labels=rev(paste(RNAseq$filt$design$volunteer, 
                                                RNAseq$filt$design$time, 
                                                sep = '_'))) +
  scale_fill_gradientn(colours = rev(colPals$inferno)) +
  xlab('') +
  ylab('') +
  ggtitle('Normalized & batch-corrected') +
  theme_custom(
    base_size = 6,
    axis.text.x.bottom = element_text(angle = 90, hjust = 1, vjust = 0.3),
    legend.position = "right",
    legend.justification = "top",
    plot.title = element_text(size=16, face='bold', hjust=0.5)
  )

p1 + p2

```


## Variance partition

### Without batch-correction

```{r}
varPart <- fitExtractVarPartModel(as.matrix(RNAseq$filt$DESeq_vst),
                                  ~(1|batch) +(1|volunteer) + (1|time),
                                  RNAseq$filt$design)

vp <- sortCols( varPart ) %>% 
  as.data.frame() %>% 
  arrange(desc(`time`))

plotPercentBars(vp[1:20,]) +
  scale_fill_manual(values = colPals$factors)

```

```{r}
p3 <- plotVarPart(vp) +
  scale_fill_manual(values = colPals$factors) +
  scale_x_discrete(labels=c('Volunteer','Time','Batch','Residuals')) +
  ylab('Variance contribution (%)') + 
  labs(fill = "Factor") +
  theme_bw(base_size = 16) +
  theme(
    legend.position = "none",
    axis.title.y = element_text(size=16, face='bold'),
    panel.grid.major.y = element_line(color = "grey80", linetype = "solid", size = 1.25), 
    panel.grid.major.x = element_blank(), 
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA, size = 2),
    axis.ticks = element_line(color = "black", size = 1.25)
  )

p3

```


### With batch-correction

```{r}
varPart <- fitExtractVarPartModel(as.matrix(RNAseq$filt$DESeq_vst_nobatch),
                                  ~(1|batch) +(1|volunteer) + (1|time),
                                  RNAseq$filt$design)

vp <- sortCols( varPart ) %>% 
  as.data.frame() %>% 
  arrange(desc(`time`))

plotPercentBars(vp[1:20,]) +
  scale_fill_manual(values = colPals$factors)

```

```{r}
p4 <- plotVarPart(vp) +
  scale_x_discrete(limits = factor(names(colPals$factors), 
                                   levels = names(colPals$factors)),
                   labels = c('Volunteer','Time','Batch','Residuals')) +
  scale_fill_manual(values = colPals$factors) +
  labs(fill = "Factor") +
  theme_bw(base_size = 16) +
  theme(
    legend.position = "right",
    legend.justification = "top",
    legend.text = element_text(size = 10),
    panel.grid.major.y = element_line(color = "grey80", linetype = "solid", size = 1.25), 
    panel.grid.major.x = element_blank(), 
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA, size = 2),
    axis.ticks = element_line(color = "black", size = 1.25)
  )

p4

```

```{r, fig.height=10, fig.width=12}
p <- patchwork::wrap_plots(p1,p2,p3,p4,ncol=2)

p

```

```{r}
ggsave("plots/figS1_batch_correction.pdf", plot = p, width = 12, height = 10, units = "in", dpi = 300, device = cairo_pdf)

```


## PCA plots

### Mean

```{r, fig.height=4, fig.width=5}
# PCA with mean summarisation
pca <- doPCA(RNAseq$filt$DESeq_vst_nobatch)

df <- pca$pcs %>% 
  cbind(RNAseq$filt$design)

df2 <- df %>% 
  group_by(time) %>% 
  summarize(PC1 = mean(PC1),
            PC2 = mean(PC2),
            PC3 = mean(PC3),) %>% 
  dplyr::rename(time2 = time)

ggplot() +
  geom_point(data = df, aes(x=PC1, y=PC2, color=time), shape=16, size=3, stroke=0, alpha=0.4) +
  geom_point(data = df2, aes(x=PC1, y=PC2, fill=time2), color="black", shape=23, size=6, stroke=1.5, alpha=1) +
  ggrepel::geom_label_repel() +
  xlab(paste("PC1 (", round(pca$percentVar[1],0), "%)", sep = "")) +
  ylab(paste("PC2 (", round(pca$percentVar[2],0), "%)", sep = "")) +
  scale_color_manual(values = colPals$time) +
  scale_fill_manual(values = colPals$time) +
  ggtitle('All volunteers') +
  theme_bw(base_size = 16) +
  theme(
    legend.position = "none",
    plot.title = element_text(size=16, face='bold', hjust=0.5),
    axis.title = element_text(size=16, face='bold'),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA, size = 2),
    axis.ticks = element_line(color = "black", size = 1.25)
  )

```

```{r}
ggsave("plots/figS2_pca_all_volunteers.pdf", width = 5, height = 4, units = "in", dpi = 300, device = cairo_pdf)

```


### Median

```{r}
# PCA with median summarisation
pca <- doPCA(RNAseq$filt$DESeq_vst_nobatch)

df <- pca$pcs %>% 
  cbind(RNAseq$filt$design)

df2 <- df %>% 
  group_by(time) %>% 
  summarize(PC1 = median(PC1),
            PC2 = median(PC2),
            PC3 = median(PC3),) %>% 
  dplyr::rename(time2 = time)

ggplot() +
  geom_point(data = df, aes(x=PC1, y=PC2, color=time), shape=16, size=3, stroke=0, alpha=0.4) +
  geom_point(data = df2, aes(x=PC1, y=PC2, fill=time2), color="black", shape=23, size=6, stroke=1.5, alpha=1) +
  ggrepel::geom_label_repel() +
  theme_custom(legend.position = "right") +
  xlab(paste("PC1 (", round(pca$percentVar[1],0), "%)", sep = "")) +
  ylab(paste("PC2 (", round(pca$percentVar[2],0), "%)", sep = "")) +
  scale_color_manual(values = colPals$time) +
  scale_fill_manual(values = colPals$time) +
  geom_text(data = df,aes(x=PC1,y=PC2,label=volunteer))

```

```{r}
# PCA with median summarisation
pca <- doPCA(RNAseq$filt$DESeq_vst_nobatch)

df <- pca$pcs %>% 
  cbind(RNAseq$filt$design)

df2 <- df %>% 
  group_by(time) %>% 
  summarize(PC1 = median(PC1),
            PC2 = median(PC2),
            PC3 = median(PC3),) %>% 
  dplyr::rename(time2 = time)

ggplot() +
  geom_point(data = df, aes(x=PC1, y=PC3, color=time), shape=16, size=3, stroke=0, alpha=0.4) +
  geom_point(data = df2, aes(x=PC1, y=PC3, fill=time2), color="black", shape=23, size=6, stroke=1.5, alpha=1) +
  ggrepel::geom_label_repel() +
  theme_custom(legend.position = "right") +
  xlab(paste("PC1 (", round(pca$percentVar[1],0), "%)", sep = "")) +
  ylab(paste("PC3 (", round(pca$percentVar[3],0), "%)", sep = "")) +
  scale_color_manual(values = colPals$time) +
  scale_fill_manual(values = colPals$time) +
  geom_text(data = df,aes(x=PC1,y=PC3,label=volunteer))

```

```{r}
# boxplot with mean summarisation PC1
pca <- doPCA(RNAseq$filt$DESeq_vst_nobatch)

df <- pca$pcs %>% 
  cbind(RNAseq$filt$design)

df2 <- df %>% 
  group_by(time) %>% 
  select(time, PC1) %>% 
  summarize_each(dplyr::funs(mean, sd, se=sd(.)/sqrt(n())), PC1) %>% 
  dplyr::rename(time2 = time)

ggplot() +
  geom_point(data = df, aes(x=time, y=PC1, color=time), shape=16, size=3, stroke=0, alpha=0.4) +
  geom_text(data = df,aes(x=time,y=PC1,label=volunteer),hjust=1)+
  geom_errorbar(data=df2, aes(x=time2, y=mean, ymin=mean-se, ymax=mean+se), width=.1, lwd=1) +
  geom_point(data = df2, aes(x=time2, y=mean, fill=time2), color="black", shape=21, size=8, stroke=1.5, alpha=1) +
  theme_custom(legend.position = "right") +
  ylab(paste("PC1 (", round(pca$percentVar[1],0), "%)", sep = "")) +
  scale_color_manual(values = colPals$time) +
  scale_fill_manual(values = colPals$time)

```

```{r}
# boxplot with median summarisation PC1
pca <- doPCA(RNAseq$filt$DESeq_vst_nobatch)

df <- pca$pcs %>% 
  cbind(RNAseq$filt$design)

df2 <- df %>% 
  group_by(time) %>% 
  select(time, PC1) %>% 
  summarize_each(dplyr::funs(median, sd, se=sd(.)/sqrt(n())), PC1) %>% 
  dplyr::rename(time2 = time)

ggplot() +
  geom_point(data = df, aes(x=time, y=PC1, color=time), shape=16, size=3, stroke=0, alpha=0.4) +
  geom_text(data = df,aes(x=time,y=PC1,label=volunteer),hjust=1)+
  geom_errorbar(data=df2, aes(x=time2, y=median, ymin=median-se, ymax=median+se), width=.1, lwd=1) +
  geom_point(data = df2, aes(x=time2, y=median, fill=time2), color="black", shape=21, size=8, stroke=1.5, alpha=1) +
  theme_custom(legend.position = "right") +
  ylab(paste("PC1 (", round(pca$percentVar[1],0), "%)", sep = "")) +
  scale_color_manual(values = colPals$time) +
  scale_fill_manual(values = colPals$time)
  
```


```{r}
#boxplot with mean summarisation PC2
pca <- doPCA(RNAseq$filt$DESeq_vst_nobatch)

df <- pca$pcs %>% 
  cbind(RNAseq$filt$design)

df2 <- df %>% 
  group_by(time) %>% 
  select(time, PC2) %>% 
  summarize_each(dplyr::funs(mean, sd, se=sd(.)/sqrt(n())), PC2) %>% 
  dplyr::rename(time2 = time)

ggplot() +
  geom_point(data = df, aes(x=time, y=PC2, color=time), shape=16, size=3, stroke=0, alpha=0.4) +
  geom_text(data = df,aes(x=time,y=PC2,label=volunteer),hjust=1)+
  geom_errorbar(data=df2, aes(x=time2, y=mean, ymin=mean-se, ymax=mean+se), width=.1, lwd=1) +
  geom_point(data = df2, aes(x=time2, y=mean, fill=time2), color="black", shape=21, size=8, stroke=1.5, alpha=1) +
  theme_custom(legend.position = "right") +
  ylab(paste("PC2 (", round(pca$percentVar[2],0), "%)", sep = "")) +
  scale_color_manual(values = colPals$time) +
  scale_fill_manual(values = colPals$time)

```


```{r}
#boxplot with median summarisation PC2
pca <- doPCA(RNAseq$filt$DESeq_vst_nobatch)

df <- pca$pcs %>% 
  cbind(RNAseq$filt$design)

df2 <- df %>% 
  group_by(time) %>% 
  select(time, PC2) %>% 
  summarize_each(dplyr::funs(median, sd, se=sd(.)/sqrt(n())), PC2) %>% 
  dplyr::rename(time2 = time)

ggplot() +
  geom_point(data = df, aes(x=time, y=PC2, color=time), shape=16, size=3, stroke=0, alpha=0.4) +
  geom_text(data = df,aes(x=time,y=PC2,label=volunteer),hjust=1)+
  geom_errorbar(data=df2, aes(x=time2, y=median, ymin=median-se, ymax=median+se), width=.1, lwd=1) +
  geom_point(data = df2, aes(x=time2, y=median, fill=time2), color="black", shape=21, size=8, stroke=1.5, alpha=1) +
  theme_custom(legend.position = "right") +
  ylab(paste("PC2 (", round(pca$percentVar[2],0), "%)", sep = "")) +
  scale_color_manual(values = colPals$time) +
  scale_fill_manual(values = colPals$time)

```


```{r}
#boxplot with mean summarisation PC3
pca <- doPCA(RNAseq$filt$DESeq_vst_nobatch)

df <- pca$pcs %>% 
  cbind(RNAseq$filt$design)

df2 <- df %>% 
  group_by(time) %>% 
  select(time, PC3) %>% 
  summarize_each(dplyr::funs(mean, sd, se=sd(.)/sqrt(n())), PC3) %>% 
  dplyr::rename(time2 = time)

ggplot() +
  geom_point(data = df, aes(x=time, y=PC3, color=time), shape=16, size=3, stroke=0, alpha=0.4) +
  geom_text(data = df,aes(x=time,y=PC3,label=volunteer),hjust=1)+
  geom_errorbar(data=df2, aes(x=time2, y=mean, ymin=mean-se, ymax=mean+se), width=.1, lwd=1) +
  geom_point(data = df2, aes(x=time2, y=mean, fill=time2), color="black", shape=21, size=8, stroke=1.5, alpha=1) +
  theme_custom(legend.position = "right") +
  ylab(paste("PC3 (", round(pca$percentVar[3],0), "%)", sep = "")) +
  scale_color_manual(values = colPals$time) +
  scale_fill_manual(values = colPals$time)

```


```{r}
#boxplot with median summarisation PC3
pca <- doPCA(RNAseq$filt$DESeq_vst_nobatch)

df <- pca$pcs %>% 
  cbind(RNAseq$filt$design)

df2 <- df %>% 
  group_by(time) %>% 
  select(time, PC3) %>% 
  summarize_each(dplyr::funs(median, sd, se=sd(.)/sqrt(n())), PC3) %>% 
  dplyr::rename(time2 = time)

ggplot() +
  geom_point(data = df, aes(x=time, y=PC3, color=time), shape=16, size=3, stroke=0, alpha=0.4) +
  geom_text(data = df,aes(x=time,y=PC3,label=volunteer),hjust=1)+
  geom_errorbar(data=df2, aes(x=time2, y=median, ymin=median-se, ymax=median+se), width=.1, lwd=1) +
  geom_point(data = df2, aes(x=time2, y=median, fill=time2), color="black", shape=21, size=8, stroke=1.5, alpha=1) +
  theme_custom(legend.position = "right") +
  ylab(paste("PC3 (", round(pca$percentVar[3],0), "%)", sep = "")) +
  scale_color_manual(values = colPals$time) +
  scale_fill_manual(values = colPals$time)

```


# Heatmap of DE gene fold change per volunteer

```{r,echo=FALSE}
# Initiate comparisons
compListBG <- list(
  c("day0","day0"),
  c("day7","day0"),
  c("day14","day0"),
  c("day21","day0"),
  c("day28","day0")
)

# Run the calculate fold change function
FCOutBg <- calcPerVolunteerFCs(log2(RNAseq$filt$DESeq_norm),RNAseq$filt$design,comparison = compListBG)
FCOutBg <- FCOutBg[!apply(FCOutBg,1,FUN=function(x) any(is.infinite(x))),]
FCOutBg[is.na(FCOutBg)] <- 0

```

```{r,fig.width=10,fig.height=10}
mark.genes <- c('CDKN1A','NR4A1','NR4A2','MYADM','IRS2', 'CD83',
                'MCL1','XYLT1','IRF2BP2','PLEKHF2','IER5','SLC35F6')

m <- FCOutBg[rownames(FCOutBg) %in% DESeq2_DEGs_filt$padj_02$GLMtime$GeneSymbol,]

volunteer <- gsub('(v\\d+) .*$', '\\1', colnames(m))
comparison <- gsub('_', ' Vs ', gsub('v\\d+ ', '', colnames(m)))

ha_top <- HeatmapAnnotation(
  Volunteer = factor(volunteer, levels = unique(volunteer)),
  Comparison = factor(comparison, levels = unique(comparison)),
  col = list(
    Volunteer = setNames(brewer.pal(length(unique(volunteer)),"Set3"),
                         nm = unique(volunteer)),
    Comparison = setNames(brewer.pal(length(unique(comparison)),"Dark2"),
                         nm = unique(comparison))
  ),
  annotation_name_gp = gpar(fontface = 'bold'),
  border = T
)

ha_right <- rowAnnotation(
  mark = anno_mark(at=which(rownames(m) %in% mark.genes),
                   labels = rownames(m)[which(rownames(m) %in% mark.genes)],
                   padding = unit(1,"mm"))
)

hc_tree <- hclust(dist(m, method = 'euclidean'), method = 'complete')
clust <- cutree(hc_tree, k=3)

p <- Heatmap(m, name = "log2FC", 
             row_split=factor(clust, levels = c(3,2,1)), cluster_row_slices = F, cluster_rows = T,
             column_title = NULL, column_split=factor(volunteer, levels = unique(volunteer)), cluster_columns = F,
             col = circlize::colorRamp2(breaks=seq(-1, 1, length.out=21),
                                        colors=colorRampPalette(c("#2166AC", "white", "#B2182B"))(21)),
             top_annotation = ha_top,
             right_annotation = ha_right,
             width = unit(170, "mm"),
             show_row_names = F, row_title = NULL, show_row_dend = T, row_dend_width=unit(10, "mm"), row_gap = unit(2, "mm"),
             show_column_names = T, column_names_gp = gpar(fontsize = 10), column_gap = unit(2, "mm"),
             border = T)

draw(p, merge_legend = T, align_heatmap_legend = "heatmap_top")

```

```{r}
pdf("plots/figS3_heatmap_GLMtime_DEGs_fdr02_all_volunteers.pdf", width = 10, height = 10)
draw(p, merge_legend = T, align_heatmap_legend = "heatmap_top")
dev.off()

```


```{r}
# add cluster information
df <- m %>% 
  as.data.frame() %>% 
  rownames_to_column(var = 'GeneSymbol') %>% 
  add_column(Cluster = recode(.$GeneSymbol, !!!clust), .after = 'GeneSymbol') %>% 
  arrange(desc(Cluster))
colnames(df) <- gsub(' ', '_', colnames(df))

head(df)

```


# Exports

```{r}
# DE gene clusters all volunteers
openxlsx::write.xlsx(df, file = "tables/dataS1_GLMtime_DEGs_fdr02_all_volunteers.xlsx", rowNames=F, overwrite=T)

# RNAseq count table excluding volunteers 9 & 10
# df <- read.table(file = 'data/rnaseq/exon_counts_all.txt',stringsAsFactors = FALSE,sep = "\t",header = TRUE)
# df <- df[,!grepl("v10",colnames(df)) & !grepl("v9",colnames(df))]
# write.table(df, file = 'data/rnaseq/rnaseq_count_mtx.tsv', quote = F, sep = '\t', col.names = T, row.names = F)

# RNAseq design table excluding volunteers 9 & 10
# df <- RNAseq$unfilt$design[!RNAseq$unfilt$design$volunteer %in% c("v9","v10"),]
# write.table(df, file = 'data/rnaseq/rnaseq_design_mtx.tsv', quote = F, sep = '\t', col.names = T, row.names = F)

```


# SessionInfo

```{r}
sessionInfo()

```
